python/ray/serve/_private/logging_utils.py

```diff
@@ -1,18 +1,60 @@
 import logging
 import os
 from typing import Optional
+import json
+import copy
 
 import ray
-from ray.serve._private.constants import DEBUG_LOG_ENV_VAR, SERVE_LOGGER_NAME
+from ray.serve._private.constants import (
+    DEBUG_LOG_ENV_VAR,
+    SERVE_LOGGER_NAME,
+    JSONIFY_LOG_MESSAGE,
+)
 
 
 LOG_FILE_FMT = "{component_name}_{component_id}.log"
-COMPONENT_LOG_FMT = (
-    "%(levelname)s %(asctime)s {component_name} {component_id} "  # noqa:E501
-)
 MESSAGE_FMT = "%(filename)s:%(lineno)d - %(message)s"
-REQUEST_ID_FMT = "%(request_id)s "
-ROUTE_FMT = "%(route)s "
+
+# Whether to jsonify the log message.
+FLAGS_JSONIFY_LOG_MESSAGE = os.environ.get(JSONIFY_LOG_MESSAGE, "0") == "1"
+
+
+class ServeJsonFormatter(logging.Formatter):
+    """Serve Logging Json Formatter
+
+    The formatter will generate the json log format on the fly
+    based on the field of record.
+    """
+
+    REQUEST_ID_FMT = "%(request_id)s"
+    ROUTE_FMT = "%(route)s"
+    APP_NAME_FMT = "%(app_name)s"
+
+    def __init__(self, component_name: str, component_id: str):
+        self.component_log_fmt = {
+            "levelname": "%(levelname)s",
+            "asctime": "%(asctime)s",
+            "component_name": component_name,
+            "component_id": component_id,
+        }
+
+    def format(self, record):
+        # generate a format string based on the record field.
+        cur_format = copy.deepcopy(self.component_log_fmt)
+        if "request_id" in record.__dict__:
+            cur_format["request_id"] = ServeJsonFormatter.REQUEST_ID_FMT
+        if "route" in record.__dict__:
+            cur_format["route"] = ServeJsonFormatter.ROUTE_FMT
+        if "app_name" in record.__dict__:
+            cur_format["app_name"] = ServeJsonFormatter.APP_NAME_FMT
```

Comment:
I don't really understand how this works, will Python magically substitute in the values for these?

Reply:
also, slight preference for `application` instead of `app_name`

Reply:
yes, it will automatically substitute the attribute values, similar to the `filename` and `lineno` etc.